source/pervasive/seq_lib.rs

```diff
@@ -363,10 +646,664 @@ impl<A> Seq<A> {
+    pub proof fn lemma_min_properties(self)
+        ensures
+            forall |x: int| self.contains(x) ==> self.min() <= x,
+            forall |i: int| 0<= i < self.len() ==> self.min() <= self[i],
+            self.len() == 0 || self.contains(self.min()),
+        decreases 
+            self.len(),
+    {
+        if self.len() <=1 {}
+        else {
+            let elt = self.drop_first().min();
+            assert(self.subrange(1,self.len() as int).contains(elt)) by {
+                self.drop_first().lemma_min_properties()
+            }
+            assert forall |i: int| 0<= i <self.len() implies self.min() <= self[i] by {
+                assert(i==0 || self[i] == self.drop_first()[i-1]);
+                assert(forall |j: int| 0<= j < self.drop_first().len() 
+                        ==> self.drop_first().min() <= self.drop_first()[j]) by {
+                    self.drop_first().lemma_min_properties()
+                }
+            }
+        }
+    }
+
+    pub open spec fn sort(self) -> Self
+    {
+        self.sort_by(|x: int,y: int| x <= y)
+    }
+
+    pub proof fn lemma_sort_ensures(self)
+        ensures
+            self.to_multiset() =~= self.sort().to_multiset(),
+            sorted_by(self.sort(), |x: int,y: int| x <= y),
+    {
+        self.lemma_sort_by_ensures(|x: int,y: int| x <= y);
+    }
+
+    /// The maximum element in a non-empty sequence is greater than or equal to
+    /// the maxima of its non-empty subsequences.
+    pub proof fn lemma_subseq_max(self, from: int, to: int)
+        requires
+            0 <= from < to <= self.len(),
+        ensures
+            self.subrange(from, to).max() <= self.max(),
+    {
+        self.lemma_max_properties();
+        self.subrange(from, to).lemma_max_properties();
+    }
+    
+    /// The minimum element in a non-empty sequence is less than or equal to
+    /// the minima of its non-empty subsequences.
+    pub proof fn lemma_subseq_min(self, from: int, to: int)
+        requires
+            0 <= from < to <= self.len(),
+        ensures
+            self.subrange(from, to).min() >= self.min(),
+    {
+        self.lemma_min_properties();
+        self.subrange(from, to).lemma_min_properties();
+    }
+}
+
+/// Helper function to aid with merge sort
+pub closed spec fn merge_sorted_with<A>(left: Seq<A>, right: Seq<A>, leq: FnSpec(A,A) -> bool) -> Seq<A>
+    recommends
+        sorted_by(left, leq),
+        sorted_by(right, leq),
+        total_ordering(leq),
+    decreases
+        left.len(), right.len()
+{
+    if left.len() == 0 {right}
+    else if right.len() == 0 {left}
+    else if leq(left.first(), right.first()){
+        Seq::<A>::empty().push(left.first()) + merge_sorted_with(left.drop_first(), right, leq)
+    } else {
+        Seq::<A>::empty().push(right.first()) + merge_sorted_with(left, right.drop_first(), leq)
+    }
+}
+
+proof fn lemma_merge_sorted_with_ensures<A>(left: Seq<A>, right: Seq<A>, leq: FnSpec(A,A) -> bool)
```

Comment:
In Dafny, I suspect this function was executable; here it's a spec.  In theory, for a spec sort, we can use something less efficient than merge sort.  I think we can leave it for now though (maybe it'll make someone's job of writing an exec version easier).